tycho-core/src/main/java/org/eclipse/tycho/p2tools/copiedfromp2/MirrorApplication.java

```diff
@@ -396,23 +399,32 @@ private IArtifactMirrorLog getLog(File location, String root) {
         return new FileMirrorLog(absolutePath, 0, root);
     }
 
+    @SuppressWarnings("unchecked")
+    private static final IQueryable<IInstallableUnit>[] EMPTY = new IQueryable[0];
+
     private IQueryable<IInstallableUnit> performResolution(IProgressMonitor monitor) throws ProvisionException {
         IProfileRegistry registry = getProfileRegistry();
         String profileId = "MirrorApplication-" + System.currentTimeMillis(); //$NON-NLS-1$
-        IProfile profile = registry.addProfile(profileId, slicingOptions.getFilter());
-        IPlanner planner = agent.getService(IPlanner.class);
-        if (planner == null)
-            throw new IllegalStateException();
-        IProfileChangeRequest pcr = planner.createChangeRequest(profile);
-        pcr.addAll(sourceIUs);
-        IProvisioningPlan plan = planner.getProvisioningPlan(pcr, null, monitor);
-        registry.removeProfile(profileId);
-        @SuppressWarnings("unchecked")
-        IQueryable<IInstallableUnit>[] arr = new IQueryable[plan.getInstallerPlan() == null ? 1 : 2];
-        arr[0] = plan.getAdditions();
-        if (plan.getInstallerPlan() != null)
-            arr[1] = plan.getInstallerPlan().getAdditions();
-        return new CompoundQueryable<>(arr);
+        List<IQueryable<IInstallableUnit>> queryables = new ArrayList<>();
+        for (TargetEnvironment environment : environments) {
+            Map<String, String> filter = new HashMap<>(slicingOptions.getFilter());
+            filter.putAll(environment.toFilterProperties());
+            IProfile profile = registry.addProfile(profileId, filter);
+            IPlanner planner = agent.getService(IPlanner.class);
+            if (planner == null) {
+                throw new IllegalStateException();
+            }
+            IProfileChangeRequest pcr = planner.createChangeRequest(profile);
+            pcr.addAll(sourceIUs);
+            IProvisioningPlan plan = planner.getProvisioningPlan(pcr, null, monitor);
+            registry.removeProfile(profileId);
+            queryables.add(plan.getAdditions());
+            IProvisioningPlan installerPlan = plan.getInstallerPlan();
+            if (installerPlan != null) {
+                queryables.add(installerPlan.getAdditions());
+            }
+        }
+        return new CompoundQueryable<>(queryables.toArray(EMPTY));
```

Comment:
As far as I can tell, this code path is actually never called by Tycho, but I added it for completes.

Reply:
@HannesWell the `MirrorApplication` is a copy from P2 (with some adjustments from Tycho) so if you think there is something in general to improve it might be worth to contribute it to P2 as well.

But if the method is not used and serves no purpose we can also remove it.

Reply:
Yes I noticed that already and I have the intention to contribute this change to P2 as well. I could then try to incooporate the other changes as well or at least provide suitable possibilities for extensions so that Tycho does not need a copy anymore.
But for now I would only do it here to have it quicker.

Reply:
> Tycho does not need a copy anymore.
> But for now I would only do it here to have it quicker.

I think it is fine to have a copy here, maintaining things at the other projects has not worked out very good (same for ApiAnalysisApplication ...) as it often required considerable amount of changes/workarounds (and discussions) for little gain.

Reply:
I suggest we finish this first and if it works as desired we'll see if I manage to contribute it back to P2.
In case that works we can then use the enhanced classes and if even later some other changes are necessary again, nothing prevents us from adding the copy again. :)